File: backend/src/main/java/com/fincoach/service/ActionPlanService.java
package com.fincoach.service;

import com.fincoach.config.AppConstants;
import com.fincoach.model.ActionPlan;
import com.fincoach.model.FinancialProfile;
import com.fincoach.repository.ActionPlanRepository;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.stereotype.Service;

import java.time.LocalDate;
import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.Optional;

import static com.fincoach.util.NumberUtils.orZero;

@Service
public class ActionPlanService {

    private static final Logger log = LoggerFactory.getLogger(ActionPlanService.class);

    private final ActionPlanRepository actionRepo;

    public ActionPlanService(ActionPlanRepository actionRepo) {
        this.actionRepo = actionRepo;
    }

    public List<ActionPlan> getActionsForUser(String userId) {
        return actionRepo.findByUserIdOrderByPriorityAscCreatedAtDesc(userId);
    }

    public ActionPlan save(ActionPlan action) {
        return actionRepo.save(action);
    }

    public Optional<ActionPlan> findById(Long id) {
        return actionRepo.findById(id);
    }

    public boolean existsById(Long id) {
        return actionRepo.existsById(id);
    }

    public void deleteById(Long id) {
        actionRepo.deleteById(id);
    }

    public Map<String, Object> buildStats(String userId) {
        List<ActionPlan> all = actionRepo.findByUserIdOrderByPriorityAscCreatedAtDesc(userId);
        long total = all.size();
        long done = all.stream().filter(a -> "TERMINE".equals(a.getStatus())).count();
        long inProgress = all.stream().filter(a -> "EN_COURS".equals(a.getStatus())).count();
        log.debug("Stats for userId={} - total={}, done={}, inProgress={}", userId, total, done, inProgress);

        Map<String, Object> stats = new LinkedHashMap<>();
        stats.put("totalActions", total);
        stats.put("completedActions", done);
        stats.put("inProgressActions", inProgress);
        stats.put("completionRate", total > 0 ? Math.round((done * 100.0 / total) * 10) / 10.0 : 0);
        return stats;
    }

    public Map<String, Object> toResponse(ActionPlan action) {
        double progressPercent = 0;
        if (action.getTargetAmount() != null && action.getTargetAmount() > 0 && action.getCurrentAmount() != null) {
            progressPercent = Math.min(100, (action.getCurrentAmount() / action.getTargetAmount()) * 100);
        }
        Map<String, Object> actionResponse = new LinkedHashMap<>();
        actionResponse.put("id", action.getId());
        actionResponse.put("userId", action.getUserId());
        actionResponse.put("title", action.getTitle());
        actionResponse.put("description", action.getDescription());
        actionResponse.put("category", action.getCategory());
        actionResponse.put("priority", action.getPriority());
        actionResponse.put("status", action.getStatus());
        actionResponse.put("targetAmount", action.getTargetAmount());
        actionResponse.put("currentAmount", action.getCurrentAmount());
        actionResponse.put("progressPercent", Math.round(progressPercent * 10.0) / 10.0);
        actionResponse.put("deadline", action.getDeadline());
        actionResponse.put("createdAt", action.getCreatedAt());
        return actionResponse;
    }

    public void generateFromProfile(FinancialProfile profile) {
        log.debug("Evaluating action plans for userId={} (savingsRate={}%, debtRatio={}%, subscriptions={}€)",
                profile.getUserId(), profile.getSavingsRate(), profile.getDebtRatio(), profile.getSubscriptions());

        double totalIncome = orZero(profile.getMonthlyIncome()) + orZero(profile.getOtherIncome());
        List<ActionPlan> newActions = new ArrayList<>();

        if (orZero(profile.getSavingsRate()) < AppConstants.ACTION_SAVINGS_TRIGGER_RATE && totalIncome > 0
                && !actionRepo.existsByUserIdAndTitleAndStatus(profile.getUserId(), "Atteindre 10% de taux d'épargne", "EN_COURS")) {
            ActionPlan action = new ActionPlan();
            action.setUserId(profile.getUserId());
            action.setTitle("Atteindre 10% de taux d'épargne");
            action.setDescription("Mettez en place un virement automatique dès réception du salaire. Objectif : épargner "
                    + (int) (totalIncome * 0.10) + " €/mois.");
            action.setCategory("EPARGNE");
            action.setPriority("HAUTE");
            action.setTargetAmount(totalIncome * 0.10 * 12);
            action.setCurrentAmount(0.0);
            action.setDeadline(LocalDate.now().plusMonths(6));
            newActions.add(action);
        }

        if (orZero(profile.getCurrentSavings()) < totalIncome * AppConstants.ACTION_EMERGENCY_FUND_MONTHS
                && !actionRepo.existsByUserIdAndTitleAndStatus(profile.getUserId(), "Constituer un fonds d'urgence (3 mois)", "EN_COURS")) {
            ActionPlan action = new ActionPlan();
            action.setUserId(profile.getUserId());
            action.setTitle("Constituer un fonds d'urgence (3 mois)");
            action.setDescription("Objectif : " + (int) (totalIncome * AppConstants.ACTION_EMERGENCY_FUND_MONTHS) + " € sur Livret A. Ce coussin vous protège des imprévus.");
            action.setCategory("EPARGNE");
            action.setPriority("HAUTE");
            action.setTargetAmount(totalIncome * AppConstants.ACTION_EMERGENCY_FUND_MONTHS);
            action.setCurrentAmount(orZero(profile.getCurrentSavings()));
            action.setDeadline(LocalDate.now().plusMonths(12));
            newActions.add(action);
        }

        if (orZero(profile.getDebtRatio()) > AppConstants.ACTION_DEBT_RATIO_TRIGGER
                && !actionRepo.existsByUserIdAndTitleAndStatus(profile.getUserId(), "Réduire le ratio d'endettement", "EN_COURS")) {
            ActionPlan action = new ActionPlan();
            action.setUserId(profile.getUserId());
            action.setTitle("Réduire le ratio d'endettement");
            action.setDescription("Votre taux d'endettement dépasse 25%. Appliquez la méthode avalanche : remboursez en priorité la dette au taux le plus élevé.");
            action.setCategory("DETTE");
            action.setPriority("HAUTE");
            action.setTargetAmount(orZero(profile.getTotalDebt()) * 0.3);
            action.setCurrentAmount(0.0);
            action.setDeadline(LocalDate.now().plusMonths(18));
            newActions.add(action);
        }

        if (orZero(profile.getSubscriptions()) > AppConstants.ACTION_SUBSCRIPTIONS_THRESHOLD
                && !actionRepo.existsByUserIdAndTitleAndStatus(profile.getUserId(), "Auditer vos abonnements", "EN_COURS")) {
            ActionPlan action = new ActionPlan();
            action.setUserId(profile.getUserId());
            action.setTitle("Auditer vos abonnements");
            action.setDescription("Listez tous vos abonnements et supprimez ceux que vous utilisez peu. Potentiel d'économie : "
                    + (int) (orZero(profile.getSubscriptions()) * 0.3) + " €/mois.");
            action.setCategory("BUDGET");
            action.setPriority("MOYENNE");
            action.setTargetAmount(orZero(profile.getSubscriptions()) * 0.3 * 12);
            action.setCurrentAmount(0.0);
            action.setDeadline(LocalDate.now().plusWeeks(2));
            newActions.add(action);
        }

        if (!newActions.isEmpty()) {
            actionRepo.saveAll(newActions);
            log.info("{} action plan(s) generated for userId={}", newActions.size(), profile.getUserId());
        } else {
            log.debug("No new action plans to generate for userId={}", profile.getUserId());
        }
    }
}

File: backend/src/services/action-plan.service.ts
import prisma from '../config/database';
import { CONSTANTS } from '../config/constants';
import { planService } from './plan.service';
import { ActionPlan, FinancialProfile, Prisma } from '@prisma/client';
import { ActionPlanPayload } from '../types';

/**
 * Action plan service — exact port of Java ActionPlanService.java.
 */
class ActionPlanService {
  private orZero(v: number | null | undefined): number {
    return v ?? 0;
  }

  async getActionsForUser(userId: string): Promise<ActionPlan[]> {
    return prisma.actionPlan.findMany({
      where: { userId },
      orderBy: [{ deadline: 'asc' }, { createdAt: 'desc' }],
    });
  }

  async findById(id: bigint): Promise<ActionPlan | null> {
    return prisma.actionPlan.findUnique({ where: { id } });
  }

  async create(userId: string, payload: ActionPlanPayload): Promise<ActionPlan> {
    const action = await prisma.actionPlan.create({
      data: {
        userId,
        title: payload.title,
        description: payload.description,
        category: payload.category,
        priority: payload.priority ?? 'MOYENNE',
        status: 'EN_COURS',
        targetAmount: payload.targetAmount,
        currentAmount: payload.currentAmount ?? 0,
        deadline: payload.deadline ? new Date(payload.deadline) : undefined,
      },
    });
    return this.autoComplete(action);
  }

  async updateStatus(
    id: bigint,
    status?: string,
    currentAmount?: number,
    explicitStatus = false,
  ): Promise<ActionPlan> {
    const existing = await prisma.actionPlan.findUniqueOrThrow({ where: { id } });

    const updatedCurrentAmount = currentAmount !== undefined ? currentAmount : this.orZero(existing.currentAmount);
    const updatedStatus = status ?? existing.status;

    const action = await prisma.actionPlan.update({
      where: { id },
      data: {
        status: updatedStatus,
        currentAmount: updatedCurrentAmount,
      },
    });

    if (!explicitStatus) {
      return this.autoComplete(action);
    }
    return action;
  }

  async deleteById(id: bigint): Promise<void> {
    await prisma.actionPlan.delete({ where: { id } });
  }

  async buildStats(userId: string): Promise<Record<string, unknown>> {
    const all = await this.getActionsForUser(userId);
    const total = all.length;
    const done = all.filter((a) => a.status === 'TERMINE').length;
    const inProgress = all.filter((a) => a.status === 'EN_COURS').length;

    return {
      totalActions: total,
      completedActions: done,
      inProgressActions: inProgress,
      completionRate: total > 0 ? Math.round((done / total) * 1000) / 10 : 0,
    };
  }

  toResponse(action: ActionPlan): Record<string, unknown> {
    let progressPercent = 0;
    if (action.targetAmount && action.targetAmount > 0 && action.currentAmount !== null) {
      progressPercent = Math.min(100, (this.orZero(action.currentAmount) / action.targetAmount) * 100);
    }

    return {
      id: action.id.toString(),
      userId: action.userId,
      title: action.title,
      description: action.description,
      category: action.category,
      priority: action.priority,
      status: action.status,
      targetAmount: action.targetAmount,
      currentAmount: action.currentAmount,
      progressPercent: Math.round(progressPercent * 10) / 10,
      deadline: action.deadline,
      createdAt: action.createdAt,
    };
  }

  /**
   * Auto-generates action plans based on the user's financial profile.
   * Each action is created only once (checked by title uniqueness per user).
   * Exact port of Java ActionPlanService.generateFromProfile().
   */
  async generateFromProfile(profile: FinancialProfile): Promise<void> {
    const totalIncome = this.orZero(profile.monthlyIncome) + this.orZero(profile.otherIncome);
    const toCreate: Prisma.ActionPlanCreateManyInput[] = [];

    const existingTitles = await prisma.actionPlan
      .findMany({ where: { userId: profile.userId }, select: { title: true } })
      .then((rows) => new Set(rows.map((r) => r.title)));

    const today = new Date();
    const addMonths = (d: Date, m: number): Date => {
      const r = new Date(d);
      r.setMonth(r.getMonth() + m);
      return r;
    };
    const addWeeks = (d: Date, w: number): Date => {
      const r = new Date(d);
      r.setDate(r.getDate() + w * 7);
      return r;
    };

    // 1. Savings rate < 10%
    if (
      this.orZero(profile.savingsRate) < CONSTANTS.ACTION_SAVINGS_TRIGGER_RATE &&
      totalIncome > 0 &&
      !existingTitles.has("Atteindre 10% de taux d'épargne")
    ) {
      toCreate.push({
        userId: profile.userId,
        title: "Atteindre 10% de taux d'épargne",
        description: `Mettez en place un virement automatique dès réception du salaire. Objectif : épargner ${Math.floor(totalIncome * 0.1)} €/mois.`,
        category: 'EPARGNE',
        priority: 'HAUTE',
        status: 'EN_COURS',
        targetAmount: totalIncome * 0.1 * 12,
        currentAmount: 0,
        deadline: addMonths(today, 6),
      });
    }

    // 2. Emergency fund < 3 months of income
    if (
      this.orZero(profile.currentSavings) < totalIncome * CONSTANTS.ACTION_EMERGENCY_FUND_MONTHS &&
      !existingTitles.has("Constituer un fonds d'urgence (3 mois)")
    ) {
      toCreate.push({
        userId: profile.userId,
        title: "Constituer un fonds d'urgence (3 mois)",
        description: `Objectif : ${Math.floor(totalIncome * CONSTANTS.ACTION_EMERGENCY_FUND_MONTHS)} € sur Livret A. Ce coussin vous protège des imprévus.`,
        category: 'EPARGNE',
        priority: 'HAUTE',
        status: 'EN_COURS',
        targetAmount: totalIncome * CONSTANTS.ACTION_EMERGENCY_FUND_MONTHS,
        currentAmount: this.orZero(profile.currentSavings),
        deadline: addMonths(today, 12),
      });
    }

    // 3. Debt ratio > 25%
    if (
      this.orZero(profile.debtRatio) > CONSTANTS.ACTION_DEBT_RATIO_TRIGGER &&
      !existingTitles.has("Réduire le ratio d'endettement")
    ) {
      toCreate.push({
        userId: profile.userId,
        title: "Réduire le ratio d'endettement",
        description:
          "Votre taux d'endettement dépasse 25%. Appliquez la méthode avalanche : remboursez en priorité la dette au taux le plus élevé.",
        category: 'DETTE',
        priority: 'HAUTE',
        status: 'EN_COURS',
        targetAmount: this.orZero(profile.totalDebt) * 0.3,
        currentAmount: 0,
        deadline: addMonths(today, 18),
      });
    }

    // 4. Subscriptions > €60/month
    if (
      this.orZero(profile.subscriptions) > CONSTANTS.ACTION_SUBSCRIPTIONS_THRESHOLD &&
      !existingTitles.has('Auditer vos abonnements')
    ) {
      toCreate.push({
        userId: profile.userId,
        title: 'Auditer vos abonnements',
        description: `Listez tous vos abonnements et supprimez ceux que vous utilisez peu. Potentiel d'économie : ${Math.floor(this.orZero(profile.subscriptions) * 0.3)} €/mois.`,
        category: 'BUDGET',
        priority: 'MOYENNE',
        status: 'EN_COURS',
        targetAmount: this.orZero(profile.subscriptions) * 0.3 * 12,
        currentAmount: 0,
        deadline: addWeeks(today, 2),
      });
    }

    if (toCreate.length > 0) {
      // Auto-generation also respects the Freemium cap (TICKET-16): never push
      // a Freemium user beyond their max simultaneous action plans.
      const slots = await planService.remainingActionPlanSlots(profile.userId);
      const capped = Number.isFinite(slots) ? toCreate.slice(0, slots) : toCreate;
      if (capped.length > 0) {
        await prisma.actionPlan.createMany({ data: capped });
        console.log(`[ActionPlanService] ${capped.length} action(s) generated for userId=${profile.userId}`);
      }
    }
  }

  // ─── Private helpers ──────────────────────────────────────────────────────

  private async autoComplete(action: ActionPlan): Promise<ActionPlan> {
    if (
      action.targetAmount &&
      action.targetAmount > 0 &&
      action.currentAmount !== null &&
      this.orZero(action.currentAmount) >= action.targetAmount &&
      action.status !== 'TERMINE'
    ) {
      return prisma.actionPlan.update({
        where: { id: action.id },
        data: { status: 'TERMINE' },
      });
    }
    return action;
  }
}

export const actionPlanService = new ActionPlanService();
